### ISSUE TITLE
fn_repl_hash_binary is not available in SQL 2005

### ISSUE THREAD

**Issue by sql-williamd** [OpenQueryStore/OpenQueryStore#16]
fn_repl_hash_binary is used to generate a query hash, however it is isn't available on SQL 2005..... wonder how we'll beat this one Enrico! :)

**Comment by Splaxi** [comment]
I don't fully understand what the fn_repl_hash_binary does, other than creating a hash that you stores.

Did you guys look into fn_varbintohexstr?

Look at the results from the below commands. Both resolves on SQL 2005 and 2008.

USE master;
GO
EXEC sp_helptext 'fn_varbintohexstr';
GO


USE master;
GO
EXEC sp_helptext 'fn_varbintohexsubstring';
GO

**Comment by Splaxi** [comment]
http://sqlblog.com/blogs/peter_larsson/archive/2010/01/27/convert-binary-value-to-string-value.aspx

**Comment by Splaxi** [comment]
http://docs.basex.org/wiki/Hashing_Module

**Comment by Evdlaar** [comment]
So to give an update on this.
We've chosen the current hashing functionality over the HASH BYTES function because the HASH BYTES can only limit a fixed amount of characters. Sadly this option is not available in 2005.

The reason we need the hashing is to uniquely identify execution plans that are related to a query statement. Through the hashing we are able to identify different plans for the same query and thus allowing an easy method to identify multiple plan usage for a single query. 

We had to resort to this method because the built-in plan hashes used in the DMVs are not trustable enough (the hash gets reused frequently). So what we basically do is shred the plan XML in the relops node that displays the operator structure of the plan. If hashes of the relop node are identical we detect an identical plan for a query, if they are different we have a new plan for the same query.

Current way of thinking is to create a custom function that performs the hashing so we can use it from 2005 and later. If there is a built-in in function that can perform the same function for a large amount of characters (4000) and is inside 2005 to 2014 we love to hear it :-).

Thanks!

**Comment by Splaxi** [comment]
I don't believe that we can find a built in function / feature that supports what you are looking for.

The best suggestion is:
http://www.databasejournal.com/features/mssql/getting-started-with-hashing-in-sql-server.html

I would argue that we should implement the recursive approach in SQL 2005 only and keep the rest of the solution it running with the current implementation.

We can use other methods / functions with the same concept of splitting the binary data into chunks and do some kind of hashing.

Here is a nvarchar(max) using CTE:
https://sqlzealots.com/2017/07/14/hashbytes-for-a-large-string-in-sql-server/

**Comment by srutzky** [comment]
This request appears to be obsolete as of changes made on Jan 29, 2018 by @Evdlaar . As you can see [in that particular commit, line 147](https://github.com/OpenQueryStore/OpenQueryStore/commit/22b43aa3c87050f7a6f0e49a5cca95dfaefa5569#diff-80a9c6e9deca8cd8e6262563c7b1f3bcL147):

```sql
SUBSTRING( [master].[sys].[fn_repl_hash_binary]( CONVERT( varbinary (MAX), [n].[query]( '.' ))), 1, 32 ),
```

was changed to:

```sql
(qs.query_hash+qs.query_plan_hash),
```

So, if OQS does not support SQL Server 2005, then it isn't due to the absence of `fn_repl_hash_binary`. If it was, then an easy solution would be to check for the SQL Server version, and if 2005, then use a different function. For example, there is the [SQL#](https://sqlsharp.com/) library (which I created) that contains a function, **Util\_HashBinary**, that does the same thing (well, **Util\_HashBinary** actually allows you to specify any of the hash algorithms available in .NET, not just `MD5` which is all that `fn_repl_hash_binary` does). And yes, it is in the Free version. People wanting that functionality in SQL Server 2005 (or to do `HASHBYTES` without the 8000 byte limit on versions prior to SQL Server 2016), might have no issues with using SQLCLR. After all, it is available on all platforms except Azure SQL Database (even if SAFE assemblies only on Linux and Azure Managed Instances).

I realize that it is preferred (for this project) to not use SQLCLR, but it's the same solution that Microsoft came up with, given that `fn_repl_hash_binary` is, in fact, SQLCLR:

```sql
SELECT [type_desc] FROM sys.all_objects WHERE [name] = N'fn_repl_hash_binary';
-- CLR_SCALAR_FUNCTION
```
